### PR TITLE
Implement Goonj Office and Coordinat0r POC  Assignment Logic For Institution

### PR DESCRIPTION
**Pull request by belwalshubham**
### Description

**Goonj Office Assignment:**
When a new institution is created, the system attempts to assign it to a relevant state office based on the provided state ID.
If no specific state office is found, the fallback office (Delhi) is assigned.

**Coordinator POC Assignment:**
The system assigns the appropriate coordinator for the institution by checking existing relationships between the state office and the institution.
If no coordinator is found, the fallback coordinator is assigned based on the institution type (Corporate, Education, etc.).

**Handling Different Institution Types:**
A map is used to determine the correct relationship type (e.g., "Corporate Coordinator of", "Family Foundation Coordinator of", etc.) based on the institution's type.

## Summary by CodeRabbit

- **New Features**
	- Introduced the `InstitutionService` class to manage institution-related operations within CiviCRM.
	- Added functionality to automatically set office and coordinator details based on specific conditions after database updates.

- **Bug Fixes**
	- Improved error handling for missing state IDs when processing institution data.

### CODE REVIEW

File: wp-content/civi-extensions/goonjcustom/Civi/InstitutionService.php [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+    $contactId = $objectRef->contact_id;
+
+    if (!$stateId) {
+      return;
+    }
+
+    if (!$stateId) {
+      \CRM_Core_Error::debug_log_message('Cannot assign Goonj Office to institution id: ' . $contactId);
+      \CRM_Core_Error::debug_log_message('No state provided on the intent for institution id: ' . $contactId);
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/Civi/InstitutionService.php:72]: Why are these two here as we are logged the contacted for both. Remove one.

File: wp-content/civi-extensions/goonjcustom/Civi/InstitutionService.php [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+      return FALSE;
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/Civi/InstitutionService.php:73]:
```suggestion
      return;
```

is not working?

File: wp-content/civi-extensions/goonjcustom/Civi/InstitutionService.php [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+    $officesFound = Contact::get(FALSE)
+      ->addSelect('id')
+      ->addWhere('contact_type', '=', 'Organization')
+      ->addWhere('contact_sub_type', 'CONTAINS', 'Goonj_Office')
+      ->addWhere('Goonj_Office_Details.Institution_Catchment', 'CONTAINS', $stateId)
+      ->execute();
+
+    $stateOffice = $officesFound->first();
+
+    error_log("stateOffice: " . print_r($stateOffice, TRUE));
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/Civi/InstitutionService.php:85]: remove it

File: wp-content/civi-extensions/goonjcustom/Civi/InstitutionService.php [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+    // If no state office is found, assign the fallback state office.
+    if (!$stateOffice) {
+      $stateOffice = self::getFallbackOffice();
+    }
+
+    $stateOfficeId = $stateOffice['id'];
+
+    Organization::update(TRUE)
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/Civi/InstitutionService.php:94]:
```suggestion
    Organization::update(FALSE)
```

File: wp-content/civi-extensions/goonjcustom/Civi/InstitutionService.php [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+   *   The parameters that were sent into the calling function.
+   */
+
+  /**
+   *
+   */
+
+  /**
+   *
+   */
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/Civi/InstitutionService.php:155]: Remove it

File: wp-content/civi-extensions/goonjcustom/Civi/InstitutionService.php [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    $coordinator = $fallbackCoordinators->itemAt($randomIndex);
+
+    return $coordinator;
+  }
+
+  /**
+   *
+   */
+  private static function getRelationshipTypeName($contactId) {
+    $organizations = Organization::get(TRUE)
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/Civi/InstitutionService.php:192]:
```suggestion
    $organizations = Organization::get(FALSE)
```

Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/Civi/InstitutionService.php:192]:
```suggestion
    $organization = Organization::get(TRUE)
```

File: wp-content/civi-extensions/goonjcustom/Civi/InstitutionService.php [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+    $firstWord = strtok($typeOfInstitution, ' ');
+
+    // Return the corresponding relationship type, or default if not found.
+    return $typeToRelationshipMap[$firstWord] ?? 'Default Coordinator of';
+  }
+
+  /**
+   *
+   */
+
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/Civi/InstitutionService.php:218]: what's this? Can you please run formatter here

File: wp-content/civi-extensions/goonjcustom/Civi/InstitutionService.php [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+    'Education Coordinator of',
+  // For Government Entity.
+    'Government Coordinator of',
+  // For Hospital.
+    'Hospital Coordinator of',
+  // For NGO.
+    'NGO Coordinator of',
+  // For Others.
+    'Default Coordinator of',
+  ];
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/Civi/InstitutionService.php:31]: Where we are using this

File: wp-content/civi-extensions/goonjcustom/Civi/InstitutionService.php [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+      $stateOffice = self::getFallbackOffice();
+    }
+
+    $stateOfficeId = $stateOffice['id'];
+
+    Organization::update(TRUE)
+      ->addValue('Review.Goonj_Office', $stateOfficeId)
+      ->addWhere('id', '=', $contactId)
+      ->execute();
+    // Get the relationship type name based on the institution type.
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/Civi/InstitutionService.php:98]: Gave one space here

**Review by tarunnjoshi — CHANGES_REQUESTED** [review]

File: wp-content/civi-extensions/goonjcustom/Civi/InstitutionService.php (hunk 7 above)
Comment (belwalshubham) [reply wp-content/civi-extensions/goonjcustom/Civi/InstitutionService.php]: @tarunnjoshi done

**Review by tarunnjoshi — APPROVED** [review]

**Pull request merged**